Support imx93 display on M33

RuoshanShi opened the pull request:
1. Added  video pll setting to support i.MX93 display. Different screens require different VIDEO PLL configurations.
2. Added rm67199 driver for i.MX93 evk board
3. Added MX8_DSI_OLED1A shield and enable the display on imx93_evk/mimx9352/m33/ddr
4. Added display tests to run the nxp_mx8_dsi_oled1a shield
5. Updated lcdifv3 driver, fixed the pixel_format setting. Added mcux_lcdifv3_set_pixel_format api.

RuoshanShi committed:
soc: imx93: enable imx93 video pll

Added video_pll instance to soc dtsi.

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi committed:
drivers: display: add RM67199 driver support

Added rm67199 driver for i.MX93 display.

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>
Signed-off-by: Winteri Wang <dongjie.wang@nxp.com>

RuoshanShi committed:
boards: nxp: add shield of NXP MX8_DSI_OLED1A for imx93_evk

Added MX8_DSI_OLED1A panel and enable display on imx93_evk ddr target

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

iuliana-prodan commented:
> I’ve been working on enabling display functionality for the i.MX93 EVK DDR target, using the MX8_DSI_OLED1A panel. Local testing has passed successfully. Your comments and suggestions are welcome!
> 
> build command : west build -b imx93_evk/mimx9352/m33/ddr --shield nxp_mx8_dsi_oled1a samples/subsys/display/lvgl -p always
> 
> <img alt="image" width="212" height="334" src="https://private-user-images.githubusercontent.com/76139036/508883461-f0067397-8b80-4931-a7c6-9b0eefd4a8ee.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIxNjQwNjksIm5iZiI6MTc2MjE2Mzc2OSwicGF0aCI6Ii83NjEzOTAzNi81MDg4ODM0NjEtZjAwNjczOTctOGI4MC00OTMxLWE3YzYtOWIwZWVmZDRhOGVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTAzVDA5NTYwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZmNDljZDA3MjhjZDU1MzdiNTlmOGE3OTk4ODUyNTZiYTUxNzgyMTY5MmMzMTg5M2E2NjRkMDVhODI1ZWIwMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nLEokG7F9AytTpd0jCmSsZF-VzaIr5HZEWi_JP6z_vQ">

And how to you load this image on m33 core?
From linux using remoteproc, from uboot, or any other way?

JiafeiPan commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts:
```diff
@@ -18,6 +19,37 @@
+	height = <1080>;
+	pixel-format = "argb-8888";
+	media-axi-clk-rate = <400000000>;
+	media-apb-clk-rate = <133333334>;
```
it seems these parameters are related with the panel to be used, so it will be more reasonable to move them to panel shield dt but not put it in board dt?

RuoshanShi replied:
> It seems these properties are duplicated with shield's dt, so remove them from board dt as they are panel related.

Maybe I can put it in dtsi and set the default value.

JiafeiPan commented on boards/shields/nxp_mx8_dsi_oled1a/shield.yml:
```diff
@@ -0,0 +1,7 @@
+  vendor: nxp
+  supported_features:
+    - display
+    - input
```
is this panel a touch screen? don't find any related dt nodes or driver.

RuoshanShi replied:
> is this panel a touch screen? don't find any related dt nodes or driver.

No, it do not support touch function.

RuoshanShi commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts:
```diff
@@ -18,6 +19,37 @@
 	ddr: memory@84000000 {
 		device_type = "memory";
-		reg = <0x84000000 DT_SIZE_M(4)>;
+		reg = <0x84000000 DT_SIZE_M(64)>;
```
> why need so much memory? I want to understand at least how much memory is used for display, thanks.
I can set it to 32M.
Memory region         Used Size  Region Size  %age Used
           FLASH:          0 GB         0 GB
             RAM:    33527696 B        32 MB     99.92%
        IDT_LIST:          0 GB        32 KB      0.00%

JarmouniA commented on drivers/display/Kconfig.rm67199:
```diff
@@ -0,0 +1,10 @@
+	select MIPI_DSI
+	depends on DT_HAS_RAYDIUM_RM67199_ENABLED
+	help
+	  Enable driver for RM67199 display.
```
```suggestion
	  Enable driver for RM67199 display controller.
```

RuoshanShi committed:
tests: drivers: Add nxp_mx8_dsi_oled1a to display tests

Updated display tests to run the nxp_mx8_dsi_oled1a shield.

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi commented on dts/bindings/display/raydium,rm67199.yaml:
```diff
@@ -0,0 +1,43 @@
+    description:
+      Number of data lanes.
+
+  pixel-format:
```
> https://github.com/zephyrproject-rtos/zephyr/blob/c9b01e8257dbe54e0f3b765f89fb69ffbafa201f/dts/bindings/mipi-dsi/mipi-dsi-device.yaml#L20

Removed the repeat code

RuoshanShi commented on drivers/display/Kconfig.rm67199:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config RM67199
+	bool "RM67199 display driver"
```
> 

Fixed

RuoshanShi commented on drivers/display/display_rm67199.c:
```diff
@@ -0,0 +1,466 @@
+
+	if (config->bl_gpio.port != NULL) {
+		return gpio_pin_set_dt(&config->bl_gpio, 1);
+	} else {
```
> 

Fixed

RuoshanShi commented on drivers/display/display_rm67199.c:
```diff
@@ -0,0 +1,466 @@
+	.set_orientation = rm67199_set_orientation,
+};
+
+#define RM67199_PANEL(id)                                                                          \
```
> replace all instances.

Fixed

RuoshanShi commented on boards/shields/nxp_mx8_dsi_oled1a/doc/index.rst:
```diff
@@ -0,0 +1,49 @@
+Overview
+********
+
+The NXP MX8 DSI OLED1A Panel is a high-resolution OLED display panel
```
> 

Fixed

RuoshanShi commented on boards/shields/nxp_mx8_dsi_oled1a/doc/index.rst:
```diff
@@ -0,0 +1,49 @@
+The NXP MX8 DSI OLED1A Panel is a high-resolution OLED display panel
+designed for use with NXP i.MX8 series processors. This panel provides
+excellent color reproduction and contrast ratio through OLED technology.
+The display connects via MIPI DSI interface and offers superior visual
```
> 

Done

RuoshanShi commented on drivers/misc/nxp_imx93_video_pll/nxp_video_pll_ctrl.c:
> This should be in SoC's or board's initialization code, does not fit the driver definition.

Thank you for your review. The VIDEO_PLL operation indeed needs to be handled at the SoC stage. I made this change because on the iMX93 we need to support multiple panels,  different shields require separate configuration of the VIDEO_PLL frequency. By implementing it as an independent clock control driver, we can configure the VIDEO_PLL frequency more flexibly.

RuoshanShi committed:
drivers: display: mcux_lcdifv3: fix pointer cast warning

- Added mcux_lcdifv3_set_pixel_format api
- Fix pointer-to-integer cast error by using uintptr_t intermediate cast
instead of direct uint64_t cast
- Fixed the pixel_format setting on waveshare_dsi_lcd shield

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

KATE-WANG-NXP commented on drivers/display/display_rm67199.c:
```diff
@@ -0,0 +1,466 @@
+	case PIXEL_FORMAT_RGB_888:
+		data->pixel_format = MIPI_DSI_PIXFMT_RGB888;
+		param = MIPI_DCS_PIXEL_FORMAT_24BIT;
+		data->bytes_per_pixel = 3;
```
pixel_format and bytes_per_pixel should be updated after is mipi_dsi_dcs_write success

KATE-WANG-NXP commented on drivers/display/Kconfig.rm67199:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2025 NXP
```
Is this commit submitted in https://github.com/zephyrproject-rtos/zephyr/pull/93134 too?

RuoshanShi commented on drivers/display/display_rm67199.c:
```diff
@@ -0,0 +1,466 @@
+	}
+
+	/* Set pixel format */
+	if (data->pixel_format == MIPI_DSI_PIXFMT_RGB888) {
```
@KATE-WANG-NXP Sure， code has been updated

JarmouniA reviewed (commented):
@RuoshanShi Sorry for the delayed follow-up https://github.com/zephyrproject-rtos/zephyr/pull/98554#discussion_r2628290312, I missed the notif for this PR.

RuoshanShi committed:
drivers: mipi dsi: Reorganize platform-specific code in dsi nxp dwc driver

Move variable declarations into platform-specific conditional blocks
and extend i.MX93 support to include M33 core. This eliminates unused
variable compilation warnings without affecting existing platform
functionality.

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

KATE-WANG-NXP commented on drivers/display/display_mcux_lcdifv3.c:
```diff
@@ -149,15 +151,48 @@ static void mcux_lcdifv3_get_capabilities(const struct device *dev,
+		dev_data->buffer_config.pixelFormat = kLCDIFV3_PixelFormatARGB8888;
+		dev_data->pixel_bytes = 4;
+		break;
+	default:
```
PIXEL_FORMAT_ABGR_8888 was newly added, you can also support kLCDIFV3_PixelFormatABGR8888 now

KATE-WANG-NXP reviewed (changes requested):
Minor changes needed

RuoshanShi commented on drivers/display/display_mcux_lcdifv3.c:
```diff
@@ -149,15 +151,48 @@ static void mcux_lcdifv3_get_capabilities(const struct device *dev,
+		return -ENOTSUP;
+	}
+
+	dev_data->buffer_config.strideBytes =
```
> I think you should align the stride then use it to calculate the fb_bytes. But where is this strideBytes used? I cannot find where it is used, seems only used to calculate the fb_bytes, if so no need for it i think

It is indeed somewhat subtle, but configuring strideBytes based on different pixel formats is necessary. It is used by LCDIFV3_SetLayerBufferConfig(base, 0, &buffer_config);, and in the NXP driver, strideBytes is required to properly configure the CTRLDESCL_3 register.”  I will fix the code to align the stride.

RuoshanShi committed:
drivers: display: Add pixel formats to align with video formats

Add support for additional pixel formats, so that the display pixel
formats are aligned with the video pixel formats:
- PANEL_PIXEL_FORMAT_BGR_888: 24-bit BGR format with 8 bits per component
- PANEL_PIXEL_FORMAT_ABGR_8888: 32-bit ABGR with alpha channel
- PANEL_PIXEL_FORMAT_RGBA_8888: 32-bit RGBA with alpha channel
- PANEL_PIXEL_FORMAT_BGRA_8888: 32-bit BGRA with alpha channel

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi commented:
@KATE-WANG-NXP Hello Kate, the code has been updated as you suggested. Could you please help review it?

JiafeiPan commented on drivers/display/display_mcux_lcdifv3.c:
```diff
@@ -314,7 +369,7 @@ static DEVICE_API(display, mcux_lcdifv3_api) = {
 	static struct mcux_lcdifv3_data mcux_lcdifv3_data_##id = {                                 \
 		.next_idx = 0,                                                                     \
-		.pixel_bytes = GET_PIXEL_BYTES(id),                                                \
+		.pixel_format = DT_INST_PROP(id, pixel_format),                                    \
```
pixel_format is declared as `enum display_pixel_format pixel_format`; at [here](https://github.com/zephyrproject-rtos/zephyr/pull/98554/changes#diff-3fa97ec34de0321eea6977cb368dc0a5eef07b2044f22113bf2c1421206e7312R50),  but it is initialized from DT pixel-format which (per nxp,imx-lcdifv3.yaml) typically uses <zephyr/dt-bindings/display/panel.h> constants (PANEL_PIXEL_FORMAT_*), not PIXEL_FORMAT_* from include/zephyr/drivers/display.h.
And then does:
`if (data->pixel_format == PANEL_PIXEL_FORMAT_RGB_565) { ... }`
at [here ](https://github.com/zephyrproject-rtos/zephyr/pull/98554/changes#diff-3fa97ec34de0321eea6977cb368dc0a5eef07b2044f22113bf2c1421206e7312R284)

So it actually uses two type of definition defined in two header files: zephyr/dt-bindings/display/panel.h and include/zephyr/drivers/display.h for the same variable, it can't make sure they are always aligned in future although it is works now.

So suggest to store DT value in a separate `uint32_t panel_pixel_format;`, 
`.panel_pixel_format= DT_INST_PROP(id, pixel_format),`
and update [these code](https://github.com/zephyrproject-rtos/zephyr/pull/98554/changes#diff-3fa97ec34de0321eea6977cb368dc0a5eef07b2044f22113bf2c1421206e7312R283) just like the following code:

```
if (data->panel_pixel_format == PANEL_PIXEL_FORMAT_RGB_565) {
        data->pixel_format  = PIXEL_FORMAT_RGB_565;
		mcux_lcdifv3_set_pixel_format(dev, PIXEL_FORMAT_RGB_565);
} else if (data->pixel_format == PANEL_PIXEL_FORMAT_RGB_888) {
```

RuoshanShi replied:
@JiafeiPan Good catch. PANEL_PIXEL_FORMAT_* and PIXEL_FORMAT_* are related, and the relevant code has already been fixed. Please help to review
